rslib/src/import_export/text/import.rs

```diff
@@ -104,22 +113,73 @@ impl<'a> Context<'a> {
         foreign: ForeignNote,
         notetype: Arc<Notetype>,
         deck_id: DeckId,
+        dupe_resolution: DupeResolution,
     ) -> Result<LogNote> {
         let today = self.col.timing_today()?.days_elapsed;
         let (mut note, mut cards) = foreign.into_native(&notetype, deck_id, today);
-        self.import_note(&mut note, &notetype)?;
+        self.import_note(&mut note, &notetype, dupe_resolution)?;
         self.import_cards(&mut cards, note.id)?;
         self.generate_missing_cards(notetype, deck_id, &note)?;
         Ok(note.into_log_note())
     }
 
-    fn import_note(&mut self, note: &mut Note, notetype: &Notetype) -> Result<()> {
-        self.col.canonify_note_tags(note, self.usn)?;
+    /// Returns the imported notes.
+    fn import_note(
+        &mut self,
+        note: &mut Note,
+        notetype: &Notetype,
+        dupe_resolution: DupeResolution,
+    ) -> Result<()> {
         note.prepare_for_update(notetype, self.normalize_notes)?;
+
+        match dupe_resolution {
+            DupeResolution::Add => self.add_prepared_note(note),
+            DupeResolution::Ignore => match self.note_duplicate(note)? {
+                Some(_) => Ok(()),
+                None => self.add_prepared_note(note),
+            },
+            DupeResolution::Update => match self.note_duplicate(note)? {
+                Some(dupe_id) => self.update_with_prepared_note(note, dupe_id),
+                None => self.add_prepared_note(note),
+            },
+        }
+    }
+
+    /// Caller must have called [Note::prepare_for_update()].
+    fn note_duplicate(&mut self, note: &Note) -> Result<Option<NoteId>> {
+        let first_stripped = strip_html_preserving_media_filenames(&note.fields()[0]);
+        Ok(self
+            .col
+            .storage
+            .note_fields_by_checksum(note.notetype_id, note.checksum.expect("set by caller"))?
+            .into_iter()
+            .filter(|(_, field)| strip_html_preserving_media_filenames(field) == first_stripped)
+            .map(|(nid, _)| nid)
+            // FIXME: There may be multiple duplicates and the old Python code updated them all in this
+            // case. However, this is probably not generally useful and worth the additional complexity.
+            // The note should probably be skipped, or an error could be raised.
+            .next())
```

Comment:
What about .last() instead of .next()? That way we update everything in this unlikely case, but don't concern ourselves with multiple note ids.